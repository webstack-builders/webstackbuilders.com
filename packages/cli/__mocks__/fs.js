import { isString } from 'lodash'

const fs = {
  ...jest.requireActual('fs'),
}

// This is a custom function that our tests can use during setup to specify
// what the files on the "mock" filesystem should look like when any of the
// `fs` APIs are used.
let mockFiles = {}
/** @param newMockFiles - {[filepath]: contents} */
fs.__setMockFiles = (newMockFiles) => {
  mockFiles = { ...newMockFiles }
}

fs.existsSync = (path) => {
  return isString(mockFiles[path])
}

fs.mkdirSync = () => {}

fs.readFileSync = (path) => {
  return mockFiles[path]
}

fs.writeFileSync = (path, contents) => {
  mockFiles[path] = contents
}

fs.unlinkSync = (path) => {
  delete mockFiles[path]
}

module.exports = fs
