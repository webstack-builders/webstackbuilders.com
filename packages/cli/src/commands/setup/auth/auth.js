import fs from 'fs'
import path from 'path'

import execa from 'execa'
import Listr from 'listr'
import prompts from 'prompts'
import terminalLink from 'terminal-link'

import { resolveFile } from '@redwoodjs/internal'
import { getProject } from '@redwoodjs/structure'

import { getPaths, writeFilesTask } from '../../../lib'
import c from '../../../lib/colors'

const AUTH_PROVIDER_IMPORT = `import { AuthProvider } from '@redwoodjs/auth'`

const OUTPUT_PATHS = {
  auth: path.join(
    getPaths().api.lib,
    getProject().isTypeScriptProject ? 'auth.ts' : 'auth.js'
  ),
  function: path.join(
    getPaths().api.functions,
    getProject().isTypeScriptProject ? 'auth.ts' : 'auth.js'
  ),
}

const getGraphqlPath = () =>
  resolveFile(path.join(getPaths().api.functions, 'graphql'))

const getWebAppPath = () => getPaths().web.app

const getSupportedProviders = () =>
  fs
    .readdirSync(path.resolve(__dirname, 'providers'))
    .map((file) => path.basename(file, '.js'))
    .filter((file) => file !== 'README.md')

const getTemplates = () =>
  fs
    .readdirSync(path.resolve(__dirname, 'templates'))
    .reduce((templates, file) => {
      if (file === 'auth.js.template') {
        return {
          ...templates,
          base: [path.resolve(__dirname, 'templates', file)],
        }
      } else {
        const provider = file.split('.')[0]
        if (templates[provider]) {
          templates[provider].push(path.resolve(__dirname, 'templates', file))
          return { ...templates }
        } else {
          return {
            ...templates,
            [provider]: [path.resolve(__dirname, 'templates', file)],
          }
        }
      }
    }, {})

// returns the content of App.{js,tsx} with import statements added
const addWebImports = (content, imports) => {
  return `${AUTH_PROVIDER_IMPORT}\n` + imports.join('\n') + '\n' + content
}

// returns the content of App.{js,tsx} with init lines added (if there are any)
const addWebInit = (content, init) => {
  if (init) {
    return content.replace(
      'const App = () => (',
      `${init}\n\nconst App = () => (`
    )
  } else {
    return content
  }
}

const objectToComponentProps = (obj, options = { exclude: [] }) => {
  let props = []

  for (const [key, value] of Object.entries(obj)) {
    if (!options.exclude.includes(key)) {
      if (key === 'client') {
        props.push(`${key}={${value}}`)
      } else {
        props.push(`${key}="${value}"`)
      }
    }
  }

  return props
}

// returns the content of App.{js,tsx} with <AuthProvider> added
const addWebRender = (content, authProvider) => {
  const [_, indent, redwoodApolloProvider] = content.match(
    /(\s+)(<RedwoodApolloProvider>.*<\/RedwoodApolloProvider>)/s
  )

  const redwoodApolloProviderLines = redwoodApolloProvider
    .split('\n')
    .map((line) => {
      return '  ' + line
    })

  const customRenderOpen = (authProvider.render || []).reduce(
    (acc, component) => acc + indent + `<${component}>`,
    ''
  )

  const customRenderClose = (authProvider.render || []).reduce(
    (acc, component) => indent + `</${component}>` + acc,
    ''
  )

  const props = objectToComponentProps(authProvider, { exclude: ['render'] })

  const renderContent =
    customRenderOpen +
    indent +
    `<AuthProvider ${props.join(' ')}>` +
    indent +
    redwoodApolloProviderLines.join('\n') +
    indent +
    `</AuthProvider>` +
    customRenderClose

  return content.replace(
    /\s+<RedwoodApolloProvider>.*<\/RedwoodApolloProvider>/s,
    renderContent
  )
}

// returns the content of App.{js,tsx} with <AuthProvider> updated
const updateWebRender = (content, authProvider) => {
  const props = objectToComponentProps(authProvider)
  const renderContent = `<AuthProvider ${props.join(' ')}>`

  return content.replace(/<AuthProvider client={.*} type=".*">/s, renderContent)
}

// returns the content of App.{js,tsx} without the old auth import
const removeOldWebImports = (content, imports) => {
  return content.replace(`${AUTH_PROVIDER_IMPORT}\n` + imports.join('\n'), '')
}

// returns the content of App.{js,tsx} without the old auth init
const removeOldWebInit = (content, init) => {
  return content.replace(init, '')
}

// returns content with old auth provider removes
const removeOldAuthProvider = async (content) => {
  // get the current auth provider
  const [_, currentAuthProvider] = content.match(
    /<AuthProvider.*type=['"](.*)['"]/s
  )

  let oldAuthProvider
  try {
    oldAuthProvider = await import(`./providers/${currentAuthProvider}`)
  } catch (e) {
    throw new Error('Could not replace existing auth provider init')
  }

  content = removeOldWebImports(content, oldAuthProvider.config.imports)
  content = removeOldWebInit(content, oldAuthProvider.config.init)

  return content
}

// check to make sure AuthProvider doesn't exist
const checkAuthProviderExists = async () => {
  const content = fs.readFileSync(getWebAppPath()).toString()

  if (content.includes(AUTH_PROVIDER_IMPORT)) {
    throw new Error(
      'Existing auth provider found.\nUse --force to override existing provider.'
    )
  }
}

// the files to create to support auth
export const files = (provider) => {
  const templates = getTemplates()
  let files = {}

  // look for any templates for this provider
  for (const [templateProvider, templateFiles] of Object.entries(templates)) {
    if (provider === templateProvider) {
      templateFiles.forEach((templateFile) => {
        const outputPath =
          OUTPUT_PATHS[path.basename(templateFile).split('.')[1]]
        files = Object.assign(files, {
          [outputPath]: fs.readFileSync(templateFile).toString(),
        })
      })
    }
  }

  // if there are no provider-specific templates, just use the base auth one
  if (Object.keys(files).length === 0) {
    files = {
      [OUTPUT_PATHS.auth]: fs.readFileSync(templates.base[0]).toString(),
    }
  }

  return files
}

// actually inserts the required config lines into App.{js,tsx}
export const addConfigToApp = async (config, force) => {
  const webAppPath = getWebAppPath()

  let content = fs.readFileSync(webAppPath).toString()

  // update existing AuthProvider if --force else add new AuthProvider
  if (content.includes(AUTH_PROVIDER_IMPORT) && force) {
    content = await removeOldAuthProvider(content)
    content = updateWebRender(content, config.authProvider)
  } else {
    content = addWebRender(content, config.authProvider)
  }

  content = addWebImports(content, config.imports)
  content = addWebInit(content, config.init)

  fs.writeFileSync(webAppPath, content)
}

export const addApiConfig = () => {
  const graphqlPath = getGraphqlPath()

  let content = fs.readFileSync(graphqlPath).toString()

  // default to an array to avoid destructure errors
  const [_, hasAuthImport] =
    content.match(/(import {.*} from 'src\/lib\/auth.*')/s) || []

  if (!hasAuthImport) {
    // add import statement
    content = content.replace(
      /^(.*services.*)$/m,
      `$1\n\nimport { getCurrentUser } from 'src/lib/auth'`
    )
    // add object to handler
    content = content.replace(
      /^(\s*)(schema: makeMergedSchema)(.*)$/m,
      `$1getCurrentUser,\n$1$2$3`
    )
    fs.writeFileSync(graphqlPath, content)
  }
}

export const isProviderSupported = (provider) => {
  return getSupportedProviders().indexOf(provider) !== -1
}

export const apiSrcDoesExist = () => {
  return fs.existsSync(path.join(getPaths().api.src))
}

export const webIndexDoesExist = () => {
  return fs.existsSync(getWebAppPath())
}

export const graphFunctionDoesExist = () => {
  return fs.existsSync(getGraphqlPath())
}

export const command = 'auth <provider>'
export const description = 'Generate an auth configuration'
export const builder = (yargs) => {
  yargs
    .positional('provider', {
      choices: getSupportedProviders(),
      description: 'Auth provider to configure',
      type: 'string',
    })
    .option('force', {
      alias: 'f',
      default: false,
      description: 'Overwrite existing configuration',
      type: 'boolean',
    })
    .epilogue(
      `Also see the ${terminalLink(
        'Redwood CLI Reference',
        'https://redwoodjs.com/reference/command-line-interface#generate-auth'
      )}`
    )
}

export const handler = async ({ provider, force }) => {
  const providerData = await import(`./providers/${provider}`)

  // check if api/src/lib/auth.js already exists and if so, ask the user to overwrite
  if (force === false) {
    if (fs.existsSync(Object.keys(files(provider))[0])) {
      const response = await prompts({
        type: 'confirm',
        name: 'answer',
        message: `Overwrite existing ${getPaths().api.lib.replace(
          getPaths().base,
          ''
        )}/auth.[jt]s?`,
        initial: false,
      })
      force = response.answer
    }
  }

  const tasks = new Listr(
    [
      {
        title: 'Generating auth lib...',
        task: (_ctx, task) => {
          if (apiSrcDoesExist()) {
            return writeFilesTask(files(provider), {
              overwriteExisting: force,
            })
          } else {
            task.skip('api/src not found, skipping')
          }
        },
      },
      {
        title: 'Adding auth config to web...',
        task: (_ctx, task) => {
          if (webIndexDoesExist()) {
            addConfigToApp(providerData.config, force)
          } else {
            task.skip('web/src/App.{js,tsx} not found, skipping')
          }
        },
      },
      {
        title: 'Adding auth config to GraphQL API...',
        task: (_ctx, task) => {
          if (graphFunctionDoesExist()) {
            addApiConfig()
          } else {
            task.skip('GraphQL function not found, skipping')
          }
        },
      },
      {
        title: 'Adding required web packages...',
        task: async () => {
          if (!isProviderSupported(provider)) {
            throw new Error(`Unknown auth provider '${provider}'`)
          }
          await execa('yarn', [
            'workspace',
            'web',
            'add',
            ...providerData.webPackages,
            '@redwoodjs/auth',
          ])
        },
      },
      providerData.apiPackages.length > 0 && {
        title: 'Adding required api packages...',
        task: async () => {
          if (!isProviderSupported(provider)) {
            throw new Error(`Unknown auth provider '${provider}'`)
          }
          await execa('yarn', [
            'workspace',
            'api',
            'add',
            ...providerData.apiPackages,
          ])
        },
      },
      {
        title: 'Installing packages...',
        task: async () => {
          await execa('yarn', ['install'])
        },
      },
      providerData.task,
      {
        title: 'One more thing...',
        task: (_ctx, task) => {
          task.title = `One more thing...\n\n   ${providerData.notes.join(
            '\n   '
          )}\n`
        },
      },
    ].filter(Boolean),
    { collapse: false }
  )

  try {
    // Don't throw existing provider error when --force exists
    if (!force) {
      await checkAuthProviderExists()
    }

    await tasks.run()
  } catch (e) {
    console.error(c.error(e.message))
    process.exit(e?.exitCode || 1)
  }
}
