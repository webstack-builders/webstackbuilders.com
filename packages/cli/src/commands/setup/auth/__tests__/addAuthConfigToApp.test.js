import fs from 'fs'

import { addConfigToApp } from '../auth'

jest.mock('../../../../lib', () => {
  const path = require('path')
  const __dirname = path.resolve()
  return {
    getPaths: () => ({
      api: { functions: '', src: '', lib: '' },
      web: {
        src: path.join(__dirname, '../create-redwood-app/template/web/src'),
        app: path.join(
          __dirname,
          '../create-redwood-app/template/web/src/App.tsx'
        ),
      },
    }),
  }
})

// This function checks output matches
const writeFileSyncSpy = jest.fn((_, content) => {
  // Line breaks cause an issue on windows snapshots
  expect(content).toMatchSnapshot()
})

beforeEach(() => {
  jest.restoreAllMocks()
  jest.spyOn(fs, 'writeFileSync').mockImplementation(writeFileSyncSpy)
})

describe('Should add config lines to App.{js,tsx}', () => {
  it('Matches Auth0 Snapshot', async () => {
    const auth0Data = await import(`../providers/auth0`)
    await addConfigToApp(auth0Data.config, false)
  })

  it('Matches Firebase Snapshot', async () => {
    const firebaseData = await import(`../providers/firebase`)
    await addConfigToApp(firebaseData.config, false)
  })

  it('Matches netlify Snapshot', async () => {
    const netlifyData = await import(`../providers/netlify`)
    await addConfigToApp(netlifyData.config, false)
  })

  it('Matches goTrue Snapshot', async () => {
    const goTrueData = await import(`../providers/goTrue`)
    await addConfigToApp(goTrueData.config, false)
  })

  it('Matches azureActiveDirectory Snapshot', async () => {
    const azureActiveDirectoryData = await import(
      `../providers/azureActiveDirectory`
    )
    await addConfigToApp(azureActiveDirectoryData.config, false)
  })

  it('Matches ethereum Snapshot', async () => {
    const ethereumData = await import(`../providers/ethereum`)
    await addConfigToApp(ethereumData.config, false)
  })

  it('Matches magicLink Snapshot', async () => {
    const magicLinkData = await import(`../providers/magicLink`)
    await addConfigToApp(magicLinkData.config, false)
  })

  it('Matches supabase Snapshot', async () => {
    const supabaseData = await import(`../providers/supabase`)
    await addConfigToApp(supabaseData.config, false)
  })

  it('Matches nhost Snapshot', async () => {
    const nhostData = await import(`../providers/nhost`)
    await addConfigToApp(nhostData.config, false)
  })
})
