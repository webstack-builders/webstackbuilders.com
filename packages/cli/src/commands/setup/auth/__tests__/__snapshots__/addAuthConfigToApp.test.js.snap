// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should add config lines to App.{js,tsx} Matches Auth0 Snapshot 1`] = `
"import { AuthProvider } from '@redwoodjs/auth'
import { Auth0Client } from '@auth0/auth0-spa-js'
import { FatalErrorBoundary, RedwoodProvider } from '@redwoodjs/web'
import { RedwoodApolloProvider } from '@redwoodjs/web/apollo'

import FatalErrorPage from 'src/pages/FatalErrorPage'
import Routes from 'src/Routes'

import './index.css'

const auth0 = new Auth0Client({
  domain: process.env.AUTH0_DOMAIN,
  client_id: process.env.AUTH0_CLIENT_ID,
  redirect_uri: process.env.AUTH0_REDIRECT_URI,

  // ** NOTE ** Storing tokens in browser local storage provides persistence across page refreshes and browser tabs.
  // However, if an attacker can achieve running JavaScript in the SPA using a cross-site scripting (XSS) attack,
  // they can retrieve the tokens stored in local storage.
  // https://auth0.com/docs/libraries/auth0-spa-js#change-storage-options
  cacheLocation: 'localstorage',
  audience: process.env.AUTH0_AUDIENCE,

  // @MARK: useRefreshTokens is required for automatically extending sessions
  // beyond that set in the initial JWT expiration.
  //
  // @MARK: https://auth0.com/docs/tokens/refresh-tokens
  // useRefreshTokens: true,
})

const App = () => (
  <FatalErrorBoundary page={FatalErrorPage}>
    <RedwoodProvider titleTemplate=\\"%PageTitle | %AppTitle\\">
      <AuthProvider client={auth0} type=\\"auth0\\">
        <RedwoodApolloProvider>
          <Routes />
        </RedwoodApolloProvider>
      </AuthProvider>
    </RedwoodProvider>
  </FatalErrorBoundary>
)

export default App
"
`;

exports[`Should add config lines to App.{js,tsx} Matches Firebase Snapshot 1`] = `
"import { AuthProvider } from '@redwoodjs/auth'
import firebase from 'firebase/app'
import 'firebase/auth'
import { FatalErrorBoundary, RedwoodProvider } from '@redwoodjs/web'
import { RedwoodApolloProvider } from '@redwoodjs/web/apollo'

import FatalErrorPage from 'src/pages/FatalErrorPage'
import Routes from 'src/Routes'

import './index.css'

const firebaseClientConfig = {
  apiKey: process.env.FIREBASE_API_KEY,
  authDomain: process.env.FIREBASE_AUTH_DOMAIN,
  projectId: process.env.FIREBASE_PROJECT_ID,
  storageBucket: process.env.FIREBASE_STORAGE_BUCKET,
  messagingSenderId: process.env.FIREBASE_MESSAGING_SENDER_ID,
  appId: process.env.FIREBASE_APP_ID,
}

const firebaseClient = ((config) => {
  if (!firebase.apps.length) {
    firebase.initializeApp(config)
  }
  return firebase
})(firebaseClientConfig)

const App = () => (
  <FatalErrorBoundary page={FatalErrorPage}>
    <RedwoodProvider titleTemplate=\\"%PageTitle | %AppTitle\\">
      <AuthProvider client={firebaseClient} type=\\"firebase\\">
        <RedwoodApolloProvider>
          <Routes />
        </RedwoodApolloProvider>
      </AuthProvider>
    </RedwoodProvider>
  </FatalErrorBoundary>
)

export default App
"
`;

exports[`Should add config lines to App.{js,tsx} Matches azureActiveDirectory Snapshot 1`] = `
"import { AuthProvider } from '@redwoodjs/auth'
import { UserAgentApplication } from 'msal'
import { FatalErrorBoundary, RedwoodProvider } from '@redwoodjs/web'
import { RedwoodApolloProvider } from '@redwoodjs/web/apollo'

import FatalErrorPage from 'src/pages/FatalErrorPage'
import Routes from 'src/Routes'

import './index.css'

const azureActiveDirectoryClient = new UserAgentApplication({
    auth: {
      clientId: process.env.AZURE_ACTIVE_DIRECTORY_CLIENT_ID,
      authority: process.env.AZURE_ACTIVE_DIRECTORY_AUTHORITY,
      redirectUri: process.env.AZURE_ACTIVE_DIRECTORY_REDIRECT_URI,
      postLogoutRedirectUri: process.env.AZURE_ACTIVE_DIRECTORY_LOGOUT_REDIRECT_URI,
    },
  })

const App = () => (
  <FatalErrorBoundary page={FatalErrorPage}>
    <RedwoodProvider titleTemplate=\\"%PageTitle | %AppTitle\\">
      <AuthProvider client={azureActiveDirectoryClient} type=\\"azureActiveDirectory\\">
        <RedwoodApolloProvider>
          <Routes />
        </RedwoodApolloProvider>
      </AuthProvider>
    </RedwoodProvider>
  </FatalErrorBoundary>
)

export default App
"
`;

exports[`Should add config lines to App.{js,tsx} Matches ethereum Snapshot 1`] = `
"import { AuthProvider } from '@redwoodjs/auth'
import EthereumAuthClient from '@oneclickdapp/ethereum-auth'
import { ApolloClient, InMemoryCache } from '@apollo/client'
import { FetchConfigProvider, useFetchConfig } from '@redwoodjs/web'
import { FatalErrorBoundary, RedwoodProvider } from '@redwoodjs/web'
import { RedwoodApolloProvider } from '@redwoodjs/web/apollo'

import FatalErrorPage from 'src/pages/FatalErrorPage'
import Routes from 'src/Routes'

import './index.css'

let ethereum

  const ApolloInjector = ({ children }) => {
  const { uri, headers } = useFetchConfig()
  try {
    const graphQLClient = new ApolloClient({
      cache: new InMemoryCache(),
      uri,
      headers,
    })
    // Default option using Apollo Client
    const makeRequest = (mutation, variables) =>
      graphQLClient.mutate({
        mutation,
        variables,
      })

    // Alternative option using graphql-hooks
    // You'll also need to modify graphQLClient
    // const makeRequest = (query, variables) =>
    //   graphQLClient.request({
    //     query,
    //     variables,
    //   })

    ethereum = new EthereumAuthClient({
      makeRequest,
      debug: process.NODE_ENV === 'development',
    })
  } catch (e) {
    console.log(e)
  }
  return React.cloneElement(children, { client: ethereum })
}

const App = () => (
  <FatalErrorBoundary page={FatalErrorPage}>
    <RedwoodProvider titleTemplate=\\"%PageTitle | %AppTitle\\">
      <FetchConfigProvider>
      <ApolloInjector>
      <AuthProvider client={ethereum} type=\\"ethereum\\">
        <RedwoodApolloProvider>
          <Routes />
        </RedwoodApolloProvider>
      </AuthProvider>
      </ApolloInjector>
      </FetchConfigProvider>
    </RedwoodProvider>
  </FatalErrorBoundary>
)

export default App
"
`;

exports[`Should add config lines to App.{js,tsx} Matches goTrue Snapshot 1`] = `
"import { AuthProvider } from '@redwoodjs/auth'
import GoTrue from 'gotrue-js'
import { FatalErrorBoundary, RedwoodProvider } from '@redwoodjs/web'
import { RedwoodApolloProvider } from '@redwoodjs/web/apollo'

import FatalErrorPage from 'src/pages/FatalErrorPage'
import Routes from 'src/Routes'

import './index.css'

const goTrueClient = new GoTrue({
  APIUrl: 'https://MYAPP.netlify.app/.netlify/identity',
  setCookie: true,
})

const App = () => (
  <FatalErrorBoundary page={FatalErrorPage}>
    <RedwoodProvider titleTemplate=\\"%PageTitle | %AppTitle\\">
      <AuthProvider client={goTrueClient} type=\\"goTrue\\">
        <RedwoodApolloProvider>
          <Routes />
        </RedwoodApolloProvider>
      </AuthProvider>
    </RedwoodProvider>
  </FatalErrorBoundary>
)

export default App
"
`;

exports[`Should add config lines to App.{js,tsx} Matches magicLink Snapshot 1`] = `
"import { AuthProvider } from '@redwoodjs/auth'
import { Magic } from 'magic-sdk'
import { FatalErrorBoundary, RedwoodProvider } from '@redwoodjs/web'
import { RedwoodApolloProvider } from '@redwoodjs/web/apollo'

import FatalErrorPage from 'src/pages/FatalErrorPage'
import Routes from 'src/Routes'

import './index.css'

const m = new Magic(process.env.MAGICLINK_PUBLIC)

const App = () => (
  <FatalErrorBoundary page={FatalErrorPage}>
    <RedwoodProvider titleTemplate=\\"%PageTitle | %AppTitle\\">
      <AuthProvider client={m} type=\\"magicLink\\">
        <RedwoodApolloProvider>
          <Routes />
        </RedwoodApolloProvider>
      </AuthProvider>
    </RedwoodProvider>
  </FatalErrorBoundary>
)

export default App
"
`;

exports[`Should add config lines to App.{js,tsx} Matches netlify Snapshot 1`] = `
"import { AuthProvider } from '@redwoodjs/auth'
import netlifyIdentity from 'netlify-identity-widget'
import { isBrowser } from '@redwoodjs/prerender/browserUtils'
import { FatalErrorBoundary, RedwoodProvider } from '@redwoodjs/web'
import { RedwoodApolloProvider } from '@redwoodjs/web/apollo'

import FatalErrorPage from 'src/pages/FatalErrorPage'
import Routes from 'src/Routes'

import './index.css'

isBrowser && netlifyIdentity.init()

const App = () => (
  <FatalErrorBoundary page={FatalErrorPage}>
    <RedwoodProvider titleTemplate=\\"%PageTitle | %AppTitle\\">
      <AuthProvider client={netlifyIdentity} type=\\"netlify\\">
        <RedwoodApolloProvider>
          <Routes />
        </RedwoodApolloProvider>
      </AuthProvider>
    </RedwoodProvider>
  </FatalErrorBoundary>
)

export default App
"
`;

exports[`Should add config lines to App.{js,tsx} Matches nhost Snapshot 1`] = `
"import { AuthProvider } from '@redwoodjs/auth'
import { createClient } from 'nhost-js-sdk'
import { FatalErrorBoundary, RedwoodProvider } from '@redwoodjs/web'
import { RedwoodApolloProvider } from '@redwoodjs/web/apollo'

import FatalErrorPage from 'src/pages/FatalErrorPage'
import Routes from 'src/Routes'

import './index.css'

const nhostClient = createClient({
  baseURL: process.env.NHOST_BACKEND_URL,
  autoLogin: false,
});


const App = () => (
  <FatalErrorBoundary page={FatalErrorPage}>
    <RedwoodProvider titleTemplate=\\"%PageTitle | %AppTitle\\">
      <AuthProvider client={nhostClient} type=\\"nhost\\">
        <RedwoodApolloProvider>
          <Routes />
        </RedwoodApolloProvider>
      </AuthProvider>
    </RedwoodProvider>
  </FatalErrorBoundary>
)

export default App
"
`;

exports[`Should add config lines to App.{js,tsx} Matches supabase Snapshot 1`] = `
"import { AuthProvider } from '@redwoodjs/auth'
import { createClient } from '@supabase/supabase-js'
import { FatalErrorBoundary, RedwoodProvider } from '@redwoodjs/web'
import { RedwoodApolloProvider } from '@redwoodjs/web/apollo'

import FatalErrorPage from 'src/pages/FatalErrorPage'
import Routes from 'src/Routes'

import './index.css'

const supabaseClient = createClient(
    process.env.SUPABASE_URL,
    process.env.SUPABASE_KEY
  )

const App = () => (
  <FatalErrorBoundary page={FatalErrorPage}>
    <RedwoodProvider titleTemplate=\\"%PageTitle | %AppTitle\\">
      <AuthProvider client={supabaseClient} type=\\"supabase\\">
        <RedwoodApolloProvider>
          <Routes />
        </RedwoodApolloProvider>
      </AuthProvider>
    </RedwoodProvider>
  </FatalErrorBoundary>
)

export default App
"
`;
